Generate man pages (#92)
valkey-io/valkey-doc: README.md

# Valkey documentation

This repo contains the Valkey documentation in Markdown format, which is used
for generating content for the website and man pages.

## Installing man pages

This repo comes with a Makefile to build and install man pages.

    make VALKEY_ROOT=path/to/valkey
    sudo make install INSTALL_MAN_DIR=/usr/local/share/man

> Prerequisites: GNU Make, Python 3, Pandoc. Additionally, the scripts need access

Reviewer: And yaml, which is not part of the standard python 3 library. (At least as of 3.12)

Author: Oh... I didn't know. I'll add it.

to the valkey code repo, where metadata files about the commands are stored.

The pages are generated under `_build/man/` by default. The default install
location is `/usr/local/share/man` (in the appropriate subdirectories).

To uninstall the man pages, run as root `make uninstall INSTALL_MAN_DIR=/usr/local/share/man`.

It's also possible to build local HTML files for local usage, using `make html`.
These HTML files are generated under `_build/html/` by default. The starting
point of the documentation is `topics/index.html`.

## Writing docs

The content of this doc repo is backing the documentation on the website, man
pages and potentially other formats. Links between pages are *relative* and
point directly to the `.md` files as they are stored in this repo. Don't start
links with `/`. This makes sure the links point to existing files regardless of

> where in the file system they docs are located, which makes it easier find

Reviewer 2: ```suggestion
where in the file system the docs are located, which makes it easier to find
```

broken links. In text editors and in the GitHub user inteface, it's possible to
click on the links to open the corresponding Markdown page. Here's a [starting point](

Examples: `../commands/get.md` or `replication.md`.

> A few exeptions are links to the `topics/`, `commands/`, `clients/` and

Reviewer 2: ```suggestion
A few exceptions are links to the `topics/`, `commands/`, `clients/` and
```

`modules/` directories, which end with a slash. These pages are generated (with

> the exeption of `topics/` which is in `topics/index.md`).

Reviewer 2: ```suggestion
the exception of `topics/` which is in `topics/index.md`).
```

Examples: `../commands/#sorted-set`, `../topics/`, `./`.

### Topics

The files under `topics/` are generic documentation pages. The `index.md` page is a starting point.

In the top of these files, there's a frontmatter metadata section, between two

> linkes of three dashes (`---`). These are YAML fields of which we use only the

Reviewer 2: ```suggestion
lines of three dashes (`---`). These are YAML fields of which we use only the
```

`title` field (and possibly `linkTitle`). The title field is used instead of an
H1 heading in each of the pages.

### Clients, modules, libraries, tools

> We maintain links to clients, module, libraries and tool in various langauges in

Reviewer 2: ```suggestion
We maintain links to clients, modules, libraries and tools in various langauges in
```

The ``{language}`` component of the path is the path-safe representation
of the full language name which is mapped in [languages.json](./languages.json).

Each client's JSON object represents the details displayed on the [clients documentation page](https://valkey.io/clients/).

For example [clients/python/github.com/valkey-io/valkey-go.json](./clients/python/github.com/valkey-io/valkey-go.json):

```json
{
    "name": "valkey-go",
    "description": "A fast Golang Valkey client that supports Client Side Caching and Auto Pipelining.",
    "recommended": true
}
```

Modules, libraries and tools follow a similar structure under their respective directories.

### Commands

The command pages under `commands/` in this repo are not complete without some
metadata from the Valkey repo, namely the JSON files in the `src/commands/`
folder. The content of these JSON files is combined with the Markdown files in
this repo when the documentation is rendered.

See: https://github.com/valkey-io/valkey/tree/unstable/src/commands/

For each command there's a Markdown file with a complete, human-readable
description.

> We process this files to provide a better experience.

Reviewer 2: ```suggestion
We process these files to provide a better experience.
```

*   Inside text, all commands should be written in all caps, in between
    backticks.
    For example: `INCR`.

> The reply types and descriptsions are stored in separate JSON files in this doc repo.

Reviewer 2: ```suggestion
The reply types and descriptions are stored in separate JSON files in this doc repo.
```

Author: So many typos! It's my specialty. :upside_down_face:

@nil-reply: [Nil reply](https://valkey.io/topics/protocol#bulk-strings)
@null-reply: [Null reply](https://valkey.io/topics/protocol#nulls)
@boolean-reply: [Boolean reply](https://valkey.io/topics/protocol#booleans)
@double-reply: [Double reply](https://valkey.io/topics/protocol#doubles)
@big-number-reply: [Big number reply](https://valkey.io/topics/protocol#big-numbers)
@bulk-error-reply: [Bulk error reply](https://valkey.io/topics/protocol#bulk-errors)
@verbatim-string-reply: [Verbatim string reply](https://valkey.io/topics/protocol#verbatim-strings)
@map-reply: [Map reply](https://valkey.io/topics/protocol#maps)
@set-reply: [Set reply](https://valkey.io/topics/protocol#sets)
@push-reply: [Push reply](https://valkey.io/topics/protocol#pushes)
```

## Styling guidelines
### Styling guidelines

Please use the following formatting rules (aiming for smaller diffs that are easier to review):

* No need for manual lines wrapping at any specific length,
  doing so usually means that adding a word creates a cascade effect and changes other lines.
* Please avoid writing lines that are too long,
  this makes the diff harder to review when only one word is changed.

> * Start every sentence on a new line.

Reviewer: I still prefer this style for doc writing, it makes it easy to scan to the next sentence when reviewing.

Author: OK, I'll put the guideline back.

Personally, I like the Alt-Q feature in emacs (reflow the current paragraph, works also in code comments). Since I started using it, I can't stop.

Reviewer: Yeah, I know there are a bunch of different flow features on line length, but it's hard to get everyone to consistently use them. "One sentence per line" is pretty easy.

>   reviewing harding.

Reviewer 2: ```suggestion
  reviewing harder.
```

> (This section may not be up do date.)

Reviewer: It does still work, but it is generating a lot of warnings. I guess we should create an issue to update this to make sure spelling still works.

Author: With all these warnings, I'd say it doesn't work. It's not usable.

In past, it used to ignore text within backticks, but not anymore it seems.

Author: I actually tried this locally now. (Installing npm pulled down thousands of deps....)

It seems to ignore stuff within backticks, but in many cases command arguments are just bold and not within backticks. It seems possible to fix by wrapping these in backticks.

Reviewer: OK, I still think we can follow up on this separately. It would be nice to have it working.

Author: Here: #93

>     After making changes to the documentation, you can use the [spellchecker-cli package](https://www.npmjs.com/package/spellchecker-cli)

Reviewer: Why is this indented now?

Author: Mistake. Well spotted.
